src/Database/Expression/IntervalExpression.php

```diff
@@ -0,0 +1,329 @@
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.1.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Database\Expression;
+
+use Cake\Database\Exception;
+use Cake\Database\ExpressionInterface;
+use Cake\Database\ValueBinder;
+use Cake\Utility\Hash;
+use Closure;
+use DateInterval;
+use DateTimeInterface;
+
+/**
+ * An expression object to generate the SQL for an interval expression.
+ */
+class IntervalExpression implements ExpressionInterface
+{
+    /**
+     * Key mappings for DateInterval array conversion.
+     *
+     * @var array
+     */
+    protected const KEYS_REQUIRED = ['y' => 1, 'm' => 1, 'd' => 1, 'h' => 1, 'i' => 1, 's' => 1, 'f' => 1];
+
+    /**
+     * SQL unit values for DateInterval array conversion.
+     *
+     * @var string[]
+     */
+    protected const KEYS_TRANSFORM = ['YEAR', 'MONTH', 'DAY', 'HOUR', 'MINUTE', 'SECOND', 'MICROSECOND'];
+
+    /**
+     * The interval object used to construct the interval SQL statement.
+     *
+     * @var \DateInterval
+     */
+    protected $interval;
+
+    /**
+     * Overriding ExpressionInterface object for server-specific SQL.
+     *
+     * @var \Cake\Database\ExpressionInterface|null
+     */
+    protected $overrideExpression;
+
+    /**
+     * The DateTimeInterface or ExpressionInterface object used as the starting
+     * point for the interval.
+     *
+     * @var \DateTimeInterface|\Cake\Database\ExpressionInterface
+     */
+    protected $subject;
+
+    /**
+     * Array of options used to construct SQL for servers that specify
+     * intervals as an expression rather than a function.
+     *
+     * @var array
+     */
+    protected $intervalSqlOptions = [];
+
+    /**
+     * Constructor that takes the field identifier, expression, or date
+     * object as well as an interval object to use as the values to construct
+     * the interval SQL.
+     *
+     * @param mixed $fieldOrValue The value or identifier to be modified.
+     * @param \DateInterval $interval The interval value as a DateInterval object.
+     */
+    public function __construct($fieldOrValue, DateInterval $interval)
+    {
+        $this->setSubject($fieldOrValue);
+        $this->setInterval($interval)->reset();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function sql(ValueBinder $generator): string
+    {
+        $exp = $this->getOverrideExpression();
+        $sql = '';
+        if ($exp) {
+            $sql = $exp->sql($generator);
+            $this->setOverrideExpression(null);
+        } else {
+            $sql .= $this->generateIntervalSql($generator);
+        }
+        $this->reset();
+
+        return $sql;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function traverse(Closure $callable): self
```

Comment:
No typehint please on return, as this is chainable.

Reply:
@dereuromark These are fixed locally. The editor wasn't warning so I looked and found the setting was still on PS12. Thanks for your help with these!